earth2mip/networks/graphcast.py

```diff
@@ -147,32 +184,36 @@ def get_forcings(time, lat, lon):
     """
 
     forcings = xarray.Dataset()
-    forcings["datetime"] = (["batch", "time"], time)
+    # need to cast to datetime64[ns] to avoid an xarray warning
+    forcings["datetime"] = (["batch", "time"], time.astype("datetime64[ns]"))
     forcings["lon"] = (["lon"], lon)
     forcings["lat"] = (["lat"], lat)
     forcings = forcings.set_coords(["datetime", "lon", "lat"])
     seconds_since_epoch = (
         forcings.coords["datetime"].data.astype("datetime64[s]").astype(np.int64)
     )
-    t = seconds_since_epoch[..., None, None]
-    lat = lat[:, None]
-    lon = lon[None, :]
 
-    # catch this warning
-    # /usr/local/lib/python3.10/dist-packages/modulus/utils/zenith_angle.py:276:
-    # RuntimeWarning: invalid value encountered in arccos
-    # hc = np.arccos(-A / B)
-    with warnings.catch_warnings():
-        warnings.simplefilter("ignore", RuntimeWarning)
-        tisr = toa_incident_solar_radiation_accumulated(t, lat, lon)
-
-    forcings["toa_incident_solar_radiation"] = (
-        ["batch", "time", "lat", "lon"],
-        tisr,
-    )
+    # unfortunately, this needs to run on the CPU since it uses the datetime
+    # information
     add_derived_vars(forcings)
     forcings = forcings.drop_vars("datetime")
-    return forcings.transpose("batch", "time", "lat", "lon")
+    forcings = forcings.transpose("batch", "time", "lat", "lon")
+
+    # put data on same device as lat
+    if isinstance(lat, jax.Array):
+        forcings = jax.tree_map(
+            lambda x: jax.device_put(x, device=lat.device()), forcings
+        )
+        seconds_since_epoch = jax.device_put(seconds_since_epoch, device=lat.device())
+
+    forcings["toa_incident_solar_radiation"] = _get_tisr(seconds_since_epoch, lat, lon)
+
+    return forcings
+
+
+def torch_device_to_jax(device: torch.device) -> jax.Device:
+    x = torch.ones([], device=device)
```

Comment:
Some more explanation as to why this could be pretty significant. Currently, the DLPack interface with JAX doesn't expose a way to do this fully asynchronously (DLPack supports a stream argument but there isn't a way to use this meaningfully in this situation). So everytime `torch_to_jax` or `jax_to_torch` is called, everything gets synchronized with the CPU and this can become very costly if inside any critical loop. It would be a good idea to avoid this if possible.

Reply:
Understood. It  seemed like a nice trick to avoid conditional logic, but I agree there are some downsides. It's only used when graphcast is initialized so not in any inner loops. I'll leave it unchanged for now, so that I can merge this before vacay.